examples/riot/coap_channel/main.c

```diff
@@ -0,0 +1,185 @@
+//   Trigger *_triggers[1];
+// } MyReactor;
+
+// CoapChannel channel;
+
+// void MyReactor_ctor(MyReactor *self, Environment *env) {
+//   Reactor_ctor(&self->super, "MyReactor", env, NULL, NULL, 0, self->_reactions, 1, self->_triggers, 1);
+//   CoapChannel_ctor(&channel, &env, "127.0.0.1", 4444, "127.0.0.1", 5555);
+// }
+
+// /* Application */
+// ENTRY_POINT(MyReactor, FOREVER, true)
+
+// int main() {
+//   lf_start();
+//   return 0;
+// }
+
+#include "reactor-uc/reactor-uc.h"
+#include "coap_channel.h"
+
+#include <errno.h>
+#include <pthread.h>
+#include <sys/socket.h>
+#include "msg.h"
+
+#include "net/gcoap.h"
+#include "shell.h"
+
+#define REMOTE_HOST "[::1]"
+
+#define MAIN_QUEUE_SIZE (4)
+static msg_t _main_msg_queue[MAIN_QUEUE_SIZE];
+static const shell_command_t shell_commands[] = {{NULL, NULL, NULL}};
+
+typedef struct {
+  int size;
+  char msg[512];
+} lf_msg_t;
+
+size_t serialize_msg_t(const void *user_struct, size_t user_struct_size, unsigned char *msg_buf) {
+  const lf_msg_t *msg = user_struct;
+
+  memcpy(msg_buf, &msg->size, sizeof(msg->size));
+  memcpy(msg_buf + sizeof(msg->size), msg->msg, msg->size);
+
+  return sizeof(msg->size) + msg->size;
+}
+
+DEFINE_TIMER_STRUCT(Sender, t, 1)
+DEFINE_TIMER_CTOR(Sender, t, 1)
+DEFINE_REACTION_STRUCT(Sender, r, 1)
+DEFINE_REACTION_CTOR(Sender, r, 0)
+DEFINE_OUTPUT_STRUCT(Sender, out, 1)
+DEFINE_OUTPUT_CTOR(Sender, out, 1)
+
+typedef struct {
+  Reactor super;
+  TIMER_INSTANCE(Sender, t);
+  REACTION_INSTANCE(Sender, r);
+  PORT_INSTANCE(Sender, out);
+  REACTOR_BOOKKEEPING_INSTANCES(1, 1, 0);
+} Sender;
+
+DEFINE_REACTION_BODY(Sender, r) {
+  SCOPE_SELF(Sender);
+  SCOPE_ENV();
+  SCOPE_PORT(Sender, out);
+
+  printf("Timer triggered @ %" PRId64 "\n", env->get_elapsed_logical_time(env));
+  lf_msg_t val;
+  strcpy(val.msg, "Hello From Sender");
+  val.size = sizeof("Hello From Sender");
+  lf_set(out, val);
+}
+
+REACTOR_CTOR_SIGNATURE_WITH_PARAMETERS(Sender, Connection **conn_out, size_t conn_out_num) {
+  REACTOR_CTOR_PREAMBLE();
+  REACTOR_CTOR(Sender);
+  INITIALIZE_REACTION(Sender, r);
+  INITIALIZE_TIMER(Sender, t, MSEC(0), SEC(1));
+  INITIALIZE_OUTPUT(Sender, out, conn_out, conn_out_num);
+
+  TIMER_REGISTER_EFFECT(t, r);
+  REACTION_REGISTER_EFFECT(r, out);
+  OUTPUT_REGISTER_SOURCE(out, r);
+}
+
+DEFINE_FEDERATED_OUTPUT_CONNECTION(Sender, out, msg_t, 1)
+
+typedef struct {
+  FederatedConnectionBundle super;
+  CoapChannel channel;
+  FEDERATED_OUTPUT_CONNECTION_INSTANCE(Sender, out);
+  FEDERATED_CONNECTION_BUNDLE_BOOKKEEPING_INSTANCES(0, 1);
+} FEDERATED_CONNECTION_BUNDLE_NAME(Sender, Receiver);
+
+// TODO This macro does not work because I need env here.
+// FEDERATED_CONNECTION_BUNDLE_CTOR_SIGNATURE(Sender, Receiver) {
+//   FEDERATED_CONNECTION_BUNDLE_CTOR_PREAMBLE();
+//   CoapChannel_ctor(&self->channel, env, REMOTE_HOST);
+
+//   FEDERATED_CONNECTION_BUNDLE_CALL_CTOR();
+
+//   INITIALIZE_FEDERATED_OUTPUT_CONNECTION(Sender, out, serialize_msg_t);
+// }
```

Comment:
@erlingrj With the new macro system even the function name here is defined by a macro, but I need a reference to `Environment` to pass it to the new `CoapChannel`.
As a workaround I have defined the function myself without the macro.

Reply:
I will fix this